### PR TITLE
Merge HttpReceiver into Http for retry logic

### PR DESCRIPTION
**Pull request by pierreca**
# Description of the problem
The concept of Receiver in the transports in parallel of the main transport client makes implementing retry logic complicated. 
# Description of the solution
We are retiring receivers in all protocols. this is what takes care of HTTP

### CODE REVIEW

File: device/transport/http/devdoc/http_requirements.md [hunk 1]
```diff
@@ -116,26 +117,101 @@ the getReceiver method queries the client for an `HttpReceiver` object that can
...
+
+**SRS_NODE_DEVICE_HTTP_RECEIVER_16_004: [**if `opts.cron` is set it shall be a string that can be interpreted as a cron expression**]**
+
+**SRS_NODE_DEVICE_HTTP_RECEIVER_16_007: [**If `opts.cron` is not a parseable cron string, an ArgumentError should be thrown.**]**
+
+**SRS_NODE_DEVICE_HTTP_RECEIVER_16_017: [**If `opts.drain` is true all messages in the queue should be pulled at once.**]**
+
+**SRS_NODE_DEVICE_HTTP_RECEIVER_16_018: [**If `opts.drain` is false, only one message shall be received at a time**]**
+
+**SRS_NODE_DEVICE_HTTP_RECEIVER_16_019: [**If the receiver is already running with a previous configuration, the existing receiver should be restarted with the new configuration**]**
```
Comment (BertKleewein) [inline device/transport/http/devdoc/http_requirements.md]: >f the receiver is already running with a previous configuration, the existing receiver should be restarted with the new configuration [](start = 44, length = 133)

how?

Comment (BertKleewein) [reply device/transport/http/devdoc/http_requirements.md]: ah, never mind.  When  I first read this, I thought it was a requirement that you were placing on the consumer, but now I realize it's the responsibility of the receiver to restart itself.

---
In reply to: [143036839](https://github.com/Azure/azure-iot-sdk-node/pull/102#discussion_r143036839) [](ancestors = 143036839)

File: device/transport/http/src/http.ts [hunk 2]
```diff
@@ -47,20 +63,43 @@ and either:
...
-    this._http = new Base();
+    this._http = http || new Base();
+
+    this._opts = defaultOptions;
+    this._receiverStarted = false;
+
+    this.on('removeListener', () => {
+      if (this._receiverStarted && this.listeners('message').length === 0) {
+        this.disableC2D(() => {
+        debug('Http c2d message polling disabled');
```
Comment (BertKleewein) [inline device/transport/http/src/http.ts]: >debug [](start = 8, length = 5)

nit: indent

File: device/transport/http/src/http.ts [hunk 3]
```diff
@@ -319,9 +403,8 @@ export class Http extends EventEmitter implements Client.Transport, BatchingTran
    */
   /*Codes_SRS_NODE_DEVICE_HTTP_16_001: [The ‘abandon’ method shall call the ‘abandon’ method of the receiver object and pass it the message and the callback given as parameters.] */
```
Comment (BertKleewein) [inline device/transport/http/src/http.ts]: >The ‘abandon’ method shall call the ‘abandon’ method of the receiver object [](start = 40, length = 75)

WTF?  The receiver object also did feedback?  Am I confused?

File: device/transport/http/src/http.ts [hunk 4]
```diff
@@ -336,22 +419,255 @@ export class Http extends EventEmitter implements Client.Transport, BatchingTran
...
+      this._opts = defaultOptions;
+    }
+
+    /*Codes_SRS_NODE_DEVICE_HTTP_RECEIVER_16_008: [Only one of the interval, at, and cron fields should be populated: if more than one is populated, an ArgumentError shall be thrown.]*/
+    if (opts.interval && opts.cron ||
+        opts.interval && opts.at ||
+        opts.interval && opts.manualPolling ||
+        opts.at && opts.cron ||
+        opts.at && opts.manualPolling ||
+        opts.cron && opts.manualPolling) {
```
Comment (BertKleewein) [inline device/transport/http/src/http.ts]: recommend adding parentheses to make operator precedence explicit.

File: device/transport/http/src/http.ts [hunk 5]
```diff
@@ -33,6 +36,19 @@ function handleResponse(done: (err?: Error, result?: results.MessageEnqueued) =>
   };
 }
 
+let defaultOptions: HttpReceiverOptions = {
```
[the remark this answers is not in the document]
Comment (BertKleewein) [reply device/transport/http/src/http.ts]: ah, it's at the bottom.  I have to get used to that.

---
In reply to: [143037710](https://github.com/Azure/azure-iot-sdk-node/pull/102#discussion_r143037710) [](ancestors = 143037710)

File: device/transport/http/src/http.ts (hunk 3 above)
Comment (pierreca) [reply device/transport/http/src/http.ts]: yeah that was weird. not relevant anymore. i forgot to remove these requirements. fixing this.

---
In reply to: [143038637](https://github.com/Azure/azure-iot-sdk-node/pull/102#discussion_r143038637) [](ancestors = 143038637)

**Pull request merged**